### PR TITLE
Api allows only subscribers to rate articles

### PR DESCRIPTION
**Pull request by lukeperrin10**
PT story: https://www.pivotaltracker.com/story/show/178171947

## User Story 
```
As an API
In order to restrict article rating to subscribers
I need to be able to register/create subscribers and restrict rating functionality
```

## Suggested changes

- Subscriber role 
- Allow functionality to become subscriber 
- Adds stripe payment functionality
- Sets up subscription plans 
- Restricts article rating to subscribers 

## Gems/Dependencies 

- Stripe-Ruby-Mock 
- Stripe-Rails

## Lessons learned 

- How to make user creation dependent on whether stripe payment status is successful 
- How to work with subscriptions in Stripe

### CODE REVIEW

File: app/controllers/api/ratings_controller.rb [hunk 1]
```diff
@@ -1,19 +1,23 @@
...
   def create
-    article = Article.find(params[:article_id])
-    rating = Rating.find_by(user_id: current_user.id, article_id: params[:article_id])
-    if rating
-      rating.update(rating_params)
-      render_success
-    else
-      new_rating = article.ratings.create(rating_params)
-      if new_rating.persisted?
+    unless current_user.member?
```
Comment (LucasKnudsen) [inline app/controllers/api/ratings_controller.rb]: Think we can get rid of this conditional for now

File: spec/requests/api/subscriber/subscriber_is_able_to_sign_in_spec.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        expect(response).to have_http_status 200
+      end
+    end
+
+    describe 'as a journalist' do
+      before do
+        post '/api/auth/sign_in', params: {
+          email: 'mrfake@fakenews.com',
+          password: 'password',
+          source: 'public-system'
```
Comment (LucasKnudsen) [inline spec/requests/api/subscriber/subscriber_is_able_to_sign_in_spec.rb]: Deprecated

File: spec/requests/api/visitor/visitor_can_become_subscriber_spec.rb [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        role: 'subscriber'
+      }
+    end
+
+    it 'is expected to respond with status 400' do
+      expect(response).to have_http_status 400
+    end
+
+    it 'is expected to respond with a message' do
+      expect(response_json['message']).to eq 'Unable to process payment, please try again later'
```
Comment (LucasKnudsen) [inline spec/requests/api/visitor/visitor_can_become_subscriber_spec.rb:92]: error_message mebbe

File: app/controllers/api/ratings_controller.rb (hunk 1 above)
Comment (lukeperrin10) [reply app/controllers/api/ratings_controller.rb]: Fixed

File: spec/requests/api/subscriber/subscriber_is_able_to_sign_in_spec.rb (hunk 2 above)
Comment (lukeperrin10) [reply spec/requests/api/subscriber/subscriber_is_able_to_sign_in_spec.rb]: Fixed

File: app/controllers/api/registrations_controller.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
+class Api::RegistrationsController < DeviseTokenAuth::RegistrationsController
+  def create
+    customer = Stripe::Customer.list(email: params[:email]).data.first
+    customer ||= Stripe::Customer.create({ email: params[:email], source: params[:stripeToken] })
```
Comment (christerforsgren91) [inline app/controllers/api/registrations_controller.rb:4]: What is the pipes doing here?

File: spec/requests/api/subscriber/can_rate_the_article_spec.rb [hunk 5]
```diff
@@ -34,7 +34,7 @@
     end
 
     it 'is expected that member have a single rating' do
```
Comment (christerforsgren91) [inline spec/requests/api/subscriber/can_rate_the_article_spec.rb:36]: should it say subscriber and not member here? :)

File: app/controllers/api/registrations_controller.rb (hunk 4 above)
Comment (lukeperrin10) [reply app/controllers/api/registrations_controller.rb:4]: It's solar graph changing the word unless, its checking if the first customer returns a result and if not continues on with creating one.

File: spec/requests/api/subscriber/can_rate_the_article_spec.rb (hunk 5 above)
Comment (lukeperrin10) [reply spec/requests/api/subscriber/can_rate_the_article_spec.rb:36]: Good spot will fix it with the next PR.

File: app/controllers/api/registrations_controller.rb (hunk 4 above)
Comment (christerforsgren91) [reply app/controllers/api/registrations_controller.rb:4]: Allright! Was confused here 👍

File: spec/requests/api/subscriber/can_rate_the_article_spec.rb (hunk 5 above)
Comment (christerforsgren91) [reply spec/requests/api/subscriber/can_rate_the_article_spec.rb:36]: 👍

**Review by christerforsgren91 — APPROVED** [review]

**Review by LucasKnudsen — APPROVED** [review]

File: app/controllers/api/registrations_controller.rb (hunk 4 above)
Comment (offcast-dev) [reply app/controllers/api/registrations_controller.rb:4]: it is the same as typing 'unless customer' at the end of line 4. Solargraph puts those pipes. We do it to make so the line 4 is not executed if customer on line 3 if not nil

**Pull request merged**